pkg/image/image.go

```diff
@@ -59,18 +61,22 @@ func WithManifestDigest(digest string) AdditionalMetadata {
 	}
 }
 
+func WithConfig(config []byte) AdditionalMetadata {
+	return func(image *Image) error {
+		image.Metadata.RawConfig = config
+		image.Metadata.ID = fmt.Sprintf("sha256:%x", sha256.Sum256(config))
+		return nil
+	}
+}
+
 // NewImage provides a new, unread image object.
-func NewImage(image v1.Image, additionalMetadata ...AdditionalMetadata) (*Image, error) {
+func NewImage(image v1.Image, additionalMetadata ...AdditionalMetadata) *Image {
```

Comment:
No need to rehash this right now, but this API is getting a bit less intuitive (and at the same time, I don't have a strong "better" suggestion yet!). It looks like we're not using these input functions here, and we only use them on `Read`... I don't think this is obvious to callers of `NewImage` or of `Read`, perhaps we could think on this...

Reply:
IMHO the API is intuitive. Read is responsible for populating the original metadata, which means that any additional (overriding) metadata should be applied *after* the read.

I also don't think that `NewImage` should call `Read` implicitly (couldn't tell if that's what you were suggesting?), as this is an intensive operation (in my opinion constructors should rarely do IO, only prep for a future call to do IO)

Reply:
Agreed, `NewImage` shouldn't call `Read`. And the logic of making these changes after the original metadata is read makes total sense. The documentation could make it clearer what's being passed in and when it gets applied. And it could be that the mutating functions aren't needed from the consumer until the call to `Read` (up for discussion), as opposed to storing the functions on the object in between the calls to `NewImage` and `Read`.

Reply:
I see that is another way to implement this. Are there specific reasons as to why the current approach is wrong?